src/java.base/share/classes/java/lang/StringCoding.java

```diff
@@ -86,8 +88,14 @@ public static boolean hasNegatives(byte[] ba, int off, int len) {
      *   a value that is less than or equal to the index of the first negative byte
      *   in the range.
      */
-    @IntrinsicCandidate
     public static int countPositives(byte[] ba, int off, int len) {
+        Objects.requireNonNull(ba, "ba");
+        Objects.checkFromIndexSize(off, len, ba.length);
```

Comment:
I recall core libraries intentionally avoided this because of performance problems. Is it possible for us to say trust the `len` argument to be non-negative? That allows us to simplify this to `Objects.checkIndex(off, ba.length - len)`. See this usage in perf-sensitive FFM API: https://github.com/openjdk/jdk/blob/149882416a956dec728a964c150b826dd589908f/src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java#L401

Reply:
My point is this is a performance-sensitive API. We are using a known-slow check method `checkFromIndexSize` which may introduce a performance regression.

Reply:
Maybe use `jdk.internal.util.Preconditions` directly instead?
```suggestion
        Preconditions.checkFromIndexSize(off, len, ba.length, null);
```

Reply:
@liach, thanks for sharing your feedback on this draft, much appreciated. :bow: In order to address your (rightful) concerns (which I also share) regarding the performance impact, in 196fc5d, I've added `StringCodingCountPositives` JMH benchmark and produced the following results:

*OS:* Ubuntu 24.04.2
*CPU:* Intel Core Ultra 9 185H x 22

```
Mode  Cnt  Score   Error  Units
avgt   10  8.617 ± 0.646  ns/op          # 149882416a9 (strIntrinCheck)
avgt   10  8.787 ± 0.313  ns/op          # 5a1301df195 (master)
```

AFAICT, (even though the patched version appears to be performing better :stuck_out_tongue:) the scores do match. Note that I will run the entire `test/micro/org/openjdk/bench/java/lang/String*.java` suite on various platforms and make sure there are no (visible?) performance regressions, before promoting this draft to a PR.

Let me know if this addresses your concerns and if there is anything else I can do to better assess the performance impact.